### ISSUE TITLE
Mixing metrics when there are more than one instance

### ISSUE THREAD

**Issue by jmjimenez** [artprima/prometheus-metrics-bundle#38]
I have implemented 2 microservices:
- They are both using this bundle
- Each one has its one namespace (for example XXXX and YYYY)
- Both are using the same redis instance to store their metrics
- Each microservice has its own metrics entry point xxxx/metrics/prometheus and yyyy/metrics/prometheus

The problem I have found is that each entry point is listing its own metrics XXXX_http_2xx_responses_total_counter BUT it is also publishing the metrics of the other microservice YYYY_http_2xx_responses_total_counter

I can see 2 possible solutions:
- Option 1: refine redis keys to include the namespace (I think this is the proper one)
- Option 2: when gathering metrics, make sure the namespace is included in the key

**Comment by denisvmedia** [comment]
The main problem is that you are using the same Redis database for both of your services. You should avoid this. To solve your problem you can have one of the following solutions:

- Use different Redis databases
- Use a sidecar container with Redis (i.e. Redis per service instance)
- Use APCU (which I personally recommend instead of Redis)

**Comment by jmjimenez** [comment]
Thanks for your quick response!!!

I cannot use APCU because my architecture is a bit complex: each "microservice" is composed of a web server and some script listeners. If I want to collect metrics from these scripts, I cannot use APCU because as far as I know APCU is not shared between scripts and web processes

I would like to have one redis for each microservice. But this scenario will make my deployment harder (I have tried...)

In fact, **I have studied the code and I think I have found a clean solution in the current implementation**: redis adapter has an additional parameter 'prefix'. This prefix will be used to identify all keys stored and retrieved from redis. So, I can assign different prefixes to each of my microservices and it will make the work done

Thanks again for your help!!!!!
